src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -1556,106 +1523,285 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
-        private static HashSet<string> s_signatures = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
-            // Calling Add-Type
-            "Add-Type", "DllImport",
+                return null;
+            }
 
-            // Doing dynamic assembly building / method indirection
-            "DefineDynamicAssembly", "DefineDynamicModule", "DefineType", "DefineConstructor", "CreateType",
-            "DefineLiteral", "DefineEnum", "DefineField", "ILGenerator", "Emit", "UnverifiableCodeAttribute",
-            "DefinePInvokeMethod", "GetTypes", "GetAssemblies", "Methods", "Properties",
+            /// <summary>
+            /// Scan a string for suspicious content.
+            ///
+            /// This is based on the Rubin-Karp algorithm, but heavily
+            /// modified to support searching for multiple patterns at
+            /// the same time.
+            ///
+            /// The key difference from Rubin-Karp is that we don't undo
+            /// the hash of the first character as we shift along in the
+            /// input.
+            ///
+            /// Instead, we can rely on knowing we need the hashes for
+            /// shorter strings anyway, so we reuse their values in
+            /// computing the hash for the longer patterns. This lets us
+            /// use a much simpler hash as well - we can avoid the use of
+            /// mod.
+            /// </summary>
+            /// <returns>The string matching the hash, or null.</returns>
+            public static string Match(string text)
+            {
+                // The longest pattern is 29 characters.
+                // The values in the array are the computed hashes of length
+                // index-1 (so runningHash[0] holds the hash for length 1).
+                var runningHash = new uint[29];
+
+                int longestPossiblePattern = 0;
+                for (int i = 0; i < text.Length; i++)
+                {
+                    uint h = text[i];
+                    if (h >= 'A' && h <= 'Z')
+                    {
+                        h = h | 0x20; // ToLower
+                    }
 
-            // Suspicious methods / properties on "Type"
-            "GetConstructor", "GetConstructors", "GetDefaultMembers", "GetEvent", "GetEvents", "GetField",
-            "GetFields", "GetInterface", "GetInterfaceMap", "GetInterfaces", "GetMember", "GetMembers",
-            "GetMethod", "GetMethods", "GetNestedType", "GetNestedTypes", "GetProperties", "GetProperty",
-            "InvokeMember", "MakeArrayType", "MakeByRefType", "MakeGenericType", "MakePointerType",
-            "DeclaringMethod", "DeclaringType", "ReflectedType", "TypeHandle", "TypeInitializer",
-            "UnderlyingSystemType",
+                    // If the character isn't in any of our patterns,
+                    // don't bother hashing and reset the running length.
+                    if (!((h >= 'a' && h <= 'z') || h == '-'))
```

Comment:
We could use here "else if".

Reply:
Nice - I manually inlined the first `if` and didn't notice.